src/java/org/apache/cassandra/service/accord/AccordService.java

```diff
@@ -833,6 +846,143 @@ public Id nodeId()
+        if (cmdTxnState.notBlocked())
+            return null;
+        //TODO (safety): check depth
+        List<AsyncChain<Void>> chains = new ArrayList<>();
+        for (TxnId blockedBy : cmdTxnState.blockedBy)
+        {
+            if (state.knows(blockedBy)) continue;
+            // need to fetch the state
+            if (safeStore.getIfLoaded(blockedBy) != null)
+            {
+                AsyncChain<Void> chain = populate(state, safeStore, blockedBy);
+                if (chain != null)
+                    chains.add(chain);
+            }
+            else
+            {
+                // go fetch it
+                chains.add(populate(state, safeStore.commandStore(), blockedBy));
+            }
+        }
+        for (PartitionKey blockedBy : cmdTxnState.blockedByKey)
+        {
+            if (state.keys.containsKey(blockedBy)) continue;
+            if (safeStore.getCommandsForKeyIfLoaded(blockedBy) != null)
+            {
+                AsyncChain<Void> chain = populate(state, safeStore, blockedBy, txnId, safeCommand.current().executeAt());
+                if (chain != null)
+                    chains.add(chain);
+            }
+            else
+            {
+                // go fetch it
+                chains.add(populate(state, safeStore.commandStore(), blockedBy, txnId, safeCommand.current().executeAt()));
+            }
+        }
+        if (chains.isEmpty())
+            return null;
+        return AsyncChains.all(chains).map(ignore -> null);
+    }
+
+    private static AsyncChain<Void> populate(StoreTxnState.Builder state, AccordSafeCommandStore safeStore, PartitionKey pk, TxnId txnId, Timestamp executeAt)
+    {
+        AccordSafeCommandsForKey commandsForKey = safeStore.getCommandsForKeyIfLoaded(pk);
+        TxnId blocking = commandsForKey.current().blockedOnTxnId(txnId, executeAt);
+        if (blocking instanceof CommandsForKey.TxnInfo)
+            blocking = ((CommandsForKey.TxnInfo) blocking).plainTxnId();
+        state.keys.put(pk, blocking);
+        if (state.txns.containsKey(blocking)) return null;
+        if (safeStore.getIfLoaded(blocking) != null) return populate(state, safeStore, blocking);
+        return populate(state, safeStore.commandStore(), blocking);
+    }
+
+    private static StoreTxnState.TxnState populate(StoreTxnState.Builder state, Command cmd)
+    {
+        StoreTxnState.Builder.TxnBuilder cmdTxnState = state.txn(cmd.txnId(), cmd.executeAt(), cmd.saveStatus());
+        if (!cmd.hasBeen(Status.Applied) && cmd.isCommitted())
+        {
+            // check blocking state
+            Command.WaitingOn waitingOn = cmd.asCommitted().waitingOn();
+            waitingOn.waitingOn.reverseForEach(null, null, null, null, (i1, i2, i3, i4, i) -> {
+                if (i < waitingOn.txnIdCount())
+                {
+                    // blocked on txn
+                    cmdTxnState.blockedBy.add(waitingOn.txnId(i));
```

Comment:
This is going to potentially list a lot of false dependencies. We update the waitingOn bitset for transactions lazily, waiting only for the highest TxnId and hoping it will provide a cheap mechanism for notifying us about earlier TxnId without us having to load them. We should really only list one TxnId that we are blocking on, or at least we should filter the resulting list to those that are actually not Applied.

Reply:
> We should really only list one TxnId that we are blocking on, or at least we should filter the resulting list to those that are actually not Applied

can def filter for `not Applied`.  My thinking is we figure out what we think, then reify the state... we can filter while loading... this can help avoid false positives in the other directly (we only list 1 txn, then when we load it its Applied... so we are waiting on 0 txn (but actually waiting on 100)

Reply:
`AccordVirtualTables` now has the following

```
            // was it applied?  If so ignore it
            if (reason != Reason.Self && txn.saveStatus.hasBeen(Status.Applied))
                return;
```

there isn't really a reason to do this during load... if its applied we don't add deps... so just filter for the table.  We also need to also double check this isn't the txn the user asked for.